### PR TITLE
Tpetra: (matrix market) make writer safer and reader less copy-pasted

### PR DESCRIPTION
**Pull request by romintomasetti**
@trilinos/tpetra

## Motivation

* `Tpetra::MatrixMarket::Writer` wasn't complaining when the output stream could not be opened. Therefore the writer was finishing its work without actually writing anything. Now, opening the output stream is safe and throws if the output stream could not be opened (all ranks throw as for `Tpetra::MatrixMarket::Reader` using a MPI broadcasting).
* `Tpetra::MatrixMarket::Reader` was safe but had the safety code copy-pasted all over the place. Made a function to centralize that aspect.

### CODE REVIEW

File: packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp
Comment (cwpearson) [inline packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp:3866]: How about `openFileOnRankZero`?

Let's document what properties the returned `ifstream` has for the other ranks. Alternatively, return an `std::optional<std::ifstream>`

File: packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp
Comment (cwpearson) [inline packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp:6969]: similar, maybe `openFileOnRankZero`

**Review by cwpearson — CHANGES_REQUESTED** [review]
Just a few quibbles about file names and documentation

File: packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp
Comment (romintomasetti) [reply packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp:3866]: * I think using `std::optional` would lead to many additional changes, especially where the underlying value of the `std::optional<std::ifstream>` will be used... If you agree, I'll just update the documentation.
* Ok for renaming to `openFileOnRankZero`

File: packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp
Comment (romintomasetti) [reply packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp:6969]: Done :heavy_check_mark:

File: packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp
Comment (cwpearson) [reply packages/tpetra/core/inout/MatrixMarket_Tpetra.hpp:3868]: I am happy with the path of least resistance 😄

**Comment by romintomasetti** [comment]
@cwpearson Is there anything more that I should do before merging?

**Comment by cwpearson** [comment]
Apologies, I was on travel and then vacation.

**Review by cwpearson — DISMISSED** [review]

**Comment by romintomasetti** [comment]
Hi @cwpearson ! Should we add a label to go forward? Thanks!

**Comment by cwpearson** [comment]
Sorry this dropped off my radar, I will try to stay on top of it! Thanks for your help

**Comment by romintomasetti** [comment]
Hi @cwpearson ! I don't have access to the test logs, so I don't know why the tests failed. But I remember last week's issues with the rebasing of Trilinos develop branch, so I just rebased this branch on today's develop :wink: Have a nice day!

**Comment by cwpearson** [comment]
The autotester was down yesterday causing some issues with the logs. I will try again and we'll see what happens!

**Comment by romintomasetti** [comment]
@cwpearson :up: :upside_down_face:

**Comment by cwpearson** [comment]
Could you build with a few more warnings turned on? There are quite a few build errors. Alternatively, give me write access to your branch and I can try to do the final cleanup - you're looking for "allow edits from maintainers" somewhere on this branch

**Comment by romintomasetti** [comment]
@cwpearson Seems I was missing a `return in;`... My bad!

**Review by cwpearson — APPROVED** [review]

**Comment by romintomasetti** [comment]
Thanks !

**Pull request merged**
